packages/nuxt/src/pages/runtime/validate.ts

```diff
@@ -1,41 +1,29 @@
-import { createError, showError } from '#app/composables/error'
-import { useNuxtApp } from '#app/nuxt'
-import { defineNuxtRouteMiddleware, useRouter } from '#app/composables/router'
+import { createError } from '#app/composables/error'
+import { defineNuxtRouteMiddleware } from '#app/composables/router'
 
-export default defineNuxtRouteMiddleware(async (to) => {
+export default defineNuxtRouteMiddleware(async (to, from) => {
   if (!to.meta?.validate) { return }
 
-  const nuxtApp = useNuxtApp()
-  const router = useRouter()
-
   const result = await Promise.resolve(to.meta.validate(to))
   if (result === true) {
     return
   }
 
   const error = createError({
+    fatal: import.meta.client,
     statusCode: (result && result.statusCode) || 404,
     statusMessage: (result && result.statusMessage) || `Page Not Found: ${to.fullPath}`,
     data: {
       path: to.fullPath,
     },
   })
-  const unsub = router.beforeResolve((final) => {
-    unsub()
-    if (final === to) {
-      const unsub = router.afterEach(async () => {
-        unsub()
-        await nuxtApp.runWithContext(() => showError(error))
-        // We pretend to have navigated to the invalid route so
-        // that the user can return to the previous page with
-        // the back button.
-        if (typeof window !== 'undefined') {
-          window.history.pushState({}, '', to.fullPath)
-        }
-      })
-      // We stop the navigation immediately before it resolves
-      // if there is no other route matching it.
-      return false
-    }
-  })
+
+  // We pretend to have navigated to the invalid route so
+  // that the user can return to the previous page with
+  // the back button.
+  if (typeof window !== 'undefined') {
```

Comment:
Would there be any difference between using `import.meta.client` here versus `typeof window !== 'undefined'`?

Reply:
not sure where that line originally came from, but `import.meta.client` should be just as safe 👍  (they are both tree-shakeable)